design-documents/graph-ql/extensions.md

```diff
@@ -0,0 +1,112 @@
+
+#### Discussion points
+
+- If this is the route we take (i.e., extend, don't modify), do we also limit extensions from _modifying_ fields that other extensions have exposed? Or do we throw errors when there are conflicts?
+- If we allow _modification_, do we continue to call them extensions or find a better name that would more accurately represent the mutable nature of these &hellip; plugins?
+
+### Developer experience
+
+In order to provide the best developer experience, we should create and publish some simple utilities that extension developers can use, both for convenience and to keep them on the right track. To this end, [TypeScript][] will provide [the most rich editor experience](https://code.visualstudio.com/docs/editor/intellisense), even if they are using JavaScript to consume it.
+
+At the very least, extension developers should be able to extend the `typeDefs`, `resolvers` and `context` of an [Apollo Server][] configuration object. If we were to publish a `createExtension` utility function, the implementation of it could be as simple as this:
+
+```ts
+import { createExtension, gql } from 'magento-utils'
+
+export default createExtension({
+  typeDefs: gql`...`,
+  resolvers: {},
+  context: {},
+})
+```
+
+Because the utilities are written in [TypeScript][], we can generate and publish JavaScript in tandem with automatically generated [type definitions](http://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html). These definitions provide the best convenience for developers. For example, if they attempt to access the context object in a resolver, not only will they see the shape of their own context object, but it will be combined with Magento's default context object, which contains properties for things like ElasticSearch and SQL.
+
+<details>
+<summary>Click to see utility code</summary>
+
+### Magento utilities
+
+```ts
+import { GraphQLResolverMap } from 'apollo-graphql'
+export { gql } from 'apollo-server'
+import { ContextFunction, Context } from 'apollo-server-core'
+import { ExpressContext } from 'apollo-server-express/dist/ApolloServer'
+import { DocumentNode } from 'graphql'
+
+import context from '../context'
+
+/**
+ * A convenience function to ensure your extension properly implements the
+ * `Extension` interface. It also merges your `context` type with the root
+ * context and feeds it into your resolvers, again, for convenience.
+ */
+export default function createExtension<T extends {} = {}>(
+  options: Extension<T>,
+) {
+  return options
+}
+
+export interface Extension<T = object, U = T & typeof context> {
+  context?: ContextFunction<ExpressContext, Context<T>> | Context<T>
+  resolvers?: GraphQLResolverMap<Context<U>>
+  typeDefs?: DocumentNode
+}
+```
+
+</details>
+
+## Conclusion
+
+TBD
```

Comment:
I wanted to leave this open to discussion and then I can update the PR by replacing this with the outcome of those conversations.